### DISCUSSION TITLE
Portable Version?

### DISCUSSION THREAD

**Discussion by hack-wizard** [GitSquared/edex-ui#1137] [Questions]
I was curious eDEX-UI can be installed or compiled in such a way that it will run completely from a flash drive, and leave nothing on the filesystem of the computer (or at least remove them on termination).  Is this something that can be done without major recoding and if so how?

**Comment by GitSquared** [comment]
Shouldn't be too hard:
 - change electron-builder options in package.json to make a portable exec build
 - change all the hard-coded paths in src/_boot.js to relative paths so config isn't written to the computer disk's, but to the flash drive

**Reply by hack-wizard** [reply]
Thanks for the answer, I think I've got the first but by adding the following to scripts in package.json:
`"build-windows-portable": "node_modules\\.bin\\electron-builder build --win=portable",`

However, I don't have too much experience with js (I've not done any coding in over a decade) and either I'm totally blinding the hardcoded paths or just not looking in the right way/places.  Could you please tell me what lines they are on?  Also, I'd make them relative by adding a './' before them right?

**Reply by hack-wizard** [reply]
Also getting the following error on running npm run start

`(node:11872) UnhandledPromiseRejectionWarning: Error: The module '\\?\c:\Users\hack-\Downloads\edex-ui-2.2.7\node_modules\node-pty\build\Release\conpty.node' `

**Reply by GitSquared** [reply]
> Could you please tell me what lines they are on?

https://github.com/GitSquared/edex-ui/blob/c5f1948f86e6913a3c62fcfc957c3c33da231352/src/_boot.js#L46-L54
https://github.com/GitSquared/edex-ui/blob/c5f1948f86e6913a3c62fcfc957c3c33da231352/src/_renderer.js#L43-L49

They should all be in those 2 files I believe.

> Also, I'd make them relative by adding a './' before them right?

In theory yes, but in practice Electron packs source code in a zip-like archive that you'd need to get out of first. A good solution would be to leverage the CWD (current working directory), like so in JS:
```js
path.join(process.cwd(), 'yourcustom.file')
```

Notice in the code above we leverage electron's `electron.app.getPath("userData")`, which is an utility to get a [clean folder](https://github.com/GitSquared/edex-ui/wiki/userData) to write config files in.
You might be OK by just defining a new root config path at the beginning of `_boot.js` & `_renderer.js` like so:

```js
const LOCAL_CONFIG_PATH = path.join(process.cwd(), 'edex-config')
```

...and then just replacing every occurence of `electron.app.getPath("userData")` in the file with `LOCAL_CONFIG_PATH`.

---

> Also getting the following error on running npm run start

Make sure to run `npm run install-windows` (or -linux) to compile C libs used in dependencies.

Hope this helps

**Reply by hack-wizard** [reply]
Yeah, that's just hanging on me, this is the output:

![image](https://user-images.githubusercontent.com/16484287/124396289-74b6d180-dcbd-11eb-93a1-74b76ea80f74.png)

It's worth noting Visual Studio Build tools are already installed so the hanging step is probably unnecessary. is it possible for me to mod the config for this to bypass the hanging step?  If so where would that be done?

**Reply by GitSquared** [reply]
You can skip the pre-script that installs VSB tools by running `npm run install-windows --ignore-scripts`, if `npm --version` >= 7 (if not, upgrade with `npm i -g npm@latest`).

**Reply by hack-wizard** [reply]
Seemed to work, but now I'm getting this:
![image](https://user-images.githubusercontent.com/16484287/124501806-164e2980-dd77-11eb-85b9-0fc316f2cd97.png)

Any idea what I'm doing wrong or missing?

**Reply by GitSquared** [reply]
(Sorry for the delay, just started a new job and I'm a bit busy)

@hack-wizard could you share what modifications you made?

also, you don't need to run the build steps if you're starting with `npm start`

**Reply by hack-wizard** [reply]
Ok the first change was to package.json in the main dir of the stable 2.2.7 release:

`"build-windows-portable": "node_modules\\.bin\\electron-builder build --win=portable",`

Maybe that was not sufficient or I did that in the wrong place?

Also added this to _boot.js in the src dir:

```
// setting local config path for portable build,
// variable replaces all instances of electron.app.getPath("userData")
const LOCAL_CONFIG_PATH = path.join(process.cwd(), 'edex-config');
```

And as indicated in the commend used the find and replace function to replace all instances of electron.app.getPath("userData") with the variable LOCAL_CONFIG_PATH in the following code blocks:

```
const settingsFile = path.join(LOCAL_CONFIG_PATH, "settings.json");
const shortcutsFile = path.join(LOCAL_CONFIG_PATH, "shortcuts.json");
const lastWindowStateFile = path.join(LOCAL_CONFIG_PATH, "lastWindowState.json");
const themesDir = path.join(LOCAL_CONFIG_PATH, "themes");
const innerThemesDir = path.join(__dirname, "assets/themes");
const kblayoutsDir = path.join(LOCAL_CONFIG_PATH, "keyboards");
const innerKblayoutsDir = path.join(__dirname, "assets/kb_layouts");
const fontsDir = path.join(LOCAL_CONFIG_PATH, "fonts");
const innerFontsDir = path.join(__dirname, "assets/fonts");

// Unset proxy env variables to avoid connection problems on the internal websockets
// See #222
if (process.env.http_proxy) delete process.env.http_proxy;
if (process.env.https_proxy) delete process.env.https_proxy;

// Fix userData folder not setup on Windows
try {
    fs.mkdirSync(LOCAL_CONFIG_PATH);
    signale.info(`Created config dir at ${LOCAL_CONFIG_PATH}`);
} catch(e) {
    signale.info(`Base config dir is ${LOCAL_CONFIG_PATH}`);
}
// Create default settings file
if (!fs.existsSync(settingsFile)) {
    fs.writeFileSync(settingsFile, JSON.stringify({
        shell: (process.platform === "win32") ? "powershell.exe" : "bash",
        shellArgs: '',
        cwd: LOCAL_CONFIG_PATH,
        keyboard: "en-US",
        theme: "tron",
        termFontSize: 15,
        audio: true,
        audioVolume: 1.0,
        disableFeedbackAudio: false,
        clockHours: 24,
        pingAddr: "1.1.1.1",
        port: 3000,
        nointro: false,
        nocursor: false,
        forceFullscreen: true,
        allowWindowed: false,
        excludeThreadsFromToplist: true,
        hideDotfiles: false,
        fsListView: false,
        experimentalGlobeFeatures: false,
        experimentalFeatures: false
    }, "", 4));
    signale.info(`Default settings written to ${settingsFile}`);
}
```

```
// Version history logging
const versionHistoryPath = path.join(LOCAL_CONFIG_PATH, "versions_log.json");
var versionHistory = fs.existsSync(versionHistoryPath) ? require(versionHistoryPath) : {};
var version = app.getVersion();
if (typeof versionHistory[version] === "undefined") {
	versionHistory[version] = {
		firstSeen: Date.now(),
		lastSeen: Date.now()
	};
} else {
	versionHistory[version].lastSeen = Date.now();
}
```

Did I miss something or replace a path I should not have?

**Comment by chromer030** [comment]
Linux's Appimages are portable by default.

**Reply by hack-wizard** [reply]
Good to know, but most (if not all) of the computers I'd be using this on at work are Windows based.

**Reply by GitSquared** [reply]
Yes, but edex is hardcoded to write config files to ~/.config.
